include/ghex/glue/atlas/atlas_user_concepts.hpp

```diff
@@ -0,0 +1,467 @@
+                }
+
+                /** @brief multiple access get function, needed by GHEX in order to perform the packing
+                 * @tparam IterationSpace iteration space type
+                 * @param is iteration space which to loop through in order to retrieve the coordinates at which to get the data
+                 * @param buffer buffer to be filled*/
+                template <typename IterationSpace>
+                void get(const IterationSpace& is, Byte* buffer) const {
+                    for (index_t idx : is.local_index()) {
+                        for (std::size_t level = 0; level < is.levels(); ++level) {
+                            std::memcpy(buffer, &get(idx, level), sizeof(T));
+                            buffer += sizeof(T);
+                        }
+                    }
+                }
+
+                template<typename IndexContainer>
+                void pack(T* buffer, const IndexContainer& c, void*) {
+                    for (const auto& is : c) {
+                        get(is, reinterpret_cast<Byte*>(buffer));
+                    }
+                }
+
+                template<typename IndexContainer>
+                void unpack(const T* buffer, const IndexContainer& c, void*) {
+                    for (const auto& is : c) {
+                        set(is, reinterpret_cast<const Byte*>(buffer));
+                    }
+                }
+
+        };
+
+#ifdef __CUDACC__
+
+#define THREADS_PER_BLOCK 32
+
+        template <typename T, typename index_t>
+        __global__ void pack_kernel(
+                const atlas::array::ArrayView<T, 2> values,
+                const std::size_t local_index_size,
+                const index_t* local_index,
+                const std::size_t levels,
+                const std::size_t buffer_size,
+                T* buffer) {
+            auto idx = threadIdx.x + (blockIdx.x * blockDim.x);
+            if (idx < local_index_size) {
+                for(auto level = 0; level < levels; ++level) {
+                    buffer[idx * levels + level] = values(local_index[idx], level);
+                }
+            }
+        }
+
+        template <typename T, typename index_t>
+        __global__ void unpack_kernel(
+                const std::size_t buffer_size,
+                const T* buffer,
+                const std::size_t local_index_size,
+                const index_t* local_index,
+                const std::size_t levels,
+                atlas::array::ArrayView<T, 2> values) {
+            auto idx = threadIdx.x + (blockIdx.x * blockDim.x);
+            if (idx < local_index_size) {
+                for(auto level = 0; level < levels; ++level) {
+                    values(local_index[idx], level) = buffer[idx * levels + level];
+                }
+            }
+        }
+
+        /** @brief GPU data descriptor*/
+        template <typename T, typename DomainDescriptor>
+        class atlas_data_descriptor_gpu {
```

Comment:
maybe use a template parameter to specifiy the architecture

Reply:
Yep, done